pkg/scheduler/actions/backfill/backfill.go

```diff
@@ -41,78 +41,82 @@ func (alloc *backfillAction) Execute(ssn *framework.Session) {
 	glog.V(3).Infof("Enter Backfill ...")
 	defer glog.V(3).Infof("Leaving Backfill ...")
 
-	// remove the top dog tasks that are not ready to run
-	unReadyTopDogs := make(map[api.JobID]*api.JobInfo)
-	for _, node := range ssn.Nodes {
-		for _, task := range node.Tasks {
-			if task.Status != api.Allocated && task.Status != api.AllocatedOverBackfill {
-				continue
-			}
-			if _, ok := ssn.TopDogReadyJobs[task.Job]; !ok {
-				// change task status back to Pending
-				job := ssn.JobIndex[task.Job]
-				if err := job.UpdateTaskStatus(task, api.Pending); err != nil {
-					glog.Errorf("Failed to update task <%v/%v> status to %v in Session <%v>: %v",
-						task.Namespace, task.Name, api.Pending, ssn.UID, err)
-				}
-
-				unReadyTopDogs[task.Job] = ssn.JobIndex[task.Job]
+	// TODO (k82cn): When backfill, it's also need to balance between Queues.
+	for _, job := range ssn.Jobs {
+		for _, task := range job.TaskStatusIndex[api.Pending] {
+			if task.Resreq.IsEmpty() {
+				// As task did not request resources, so it only need to meet predicates.
+				// TODO (k82cn): need to prioritize nodes to avoid pod hole.
+				for _, node := range ssn.Nodes {
+					// TODO (k82cn): predicates did not consider pod number for now, there'll
+					// be ping-pong case here.
+					if err := ssn.PredicateFn(task, node); err != nil {
+						glog.V(3).Infof("Predicates failed for task <%s/%s> on node <%s>: %v",
+							task.Namespace, task.Name, node.Name, err)
+						continue
+					}
 
-				// remove the task from the node
-				node.RemoveTask(task)
-				glog.Infof("removed task %s from node, idle is %v", task.Name, node.Idle.MilliCPU)
+					glog.V(3).Infof("Binding Task <%v/%v> to node <%v>", task.Namespace, task.Name, node.Name)
+					if err := ssn.Allocate(task, node.Name, false); err != nil {
+						glog.Errorf("Failed to bind Task %v on %v in Session %v", task.UID, node.Name, ssn.UID)
+						continue
+					}
+					break
+				}
+			} else {
```

Comment:
this is already checked in. we need to merge here.

Reply:
it has been taken care of.